### PR TITLE
Cpp mcts

### PR DESCRIPTION
**Pull request by mkju19**

### CODE REVIEW

File: cpp_modules/src/mcts/Mcts.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+		node = randomChoice(maxNodes);
+		state->performAction(node->getAction());
+
+		if (node->N == 0) {
+			return {node, state};
+		}
+	}
+
+	node->expand(state);
```
Comment (selectionator) [inline cpp_modules/src/mcts/Mcts.cpp:87]: Potentially (Most likely) unreachable code

File: cpp_modules/src/mcts/Mcts.cpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+		state->performAction(action);
+	}
+
+	return state->getValue();
+}
+
+void Mcts::backPropagate(std::shared_ptr<Node> node, const int outcome) {
+	while (node != nullptr) {
+		node->N += 1;
```
Comment (selectionator) [inline cpp_modules/src/mcts/Mcts.cpp:108]: N and Q could maybe get a different name. While it does make sense, in the context of the formal definition, it's not immediately clear here what it is

File: cpp_modules/src/mcts/Mcts.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	// Check if the action matches any explored nodes
+	for (const auto childAction: _rootNode->children | std::views::keys) {
+		if (childAction == action) {
+			_rootState->performAction(action);
+			_rootNode = _rootNode->children[action];
+			return;
+		}
+	}
+
+	// If the action is not explored it will still be performed, but the root node will be reset.
```
Comment (selectionator) [inline cpp_modules/src/mcts/Mcts.cpp:134]: I prefer not being able to take an action that is not part of the child nodes

File: cpp_modules/src/mcts/Mcts.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+		static double getMaxNodeValue(const std::map<Action, std::shared_ptr<Node> > &nodes);
+
+		static std::vector<std::shared_ptr<Node> > getMaxNodes(
+			std::map<Action, std::shared_ptr<Node> > &children, double maxValue);
+
+	public:
+		std::shared_ptr<Node> getRootNode() { return _rootNode; }
+		std::shared_ptr<State> getRootState() { return _rootState; }
+
+		// std::shared_ptr<Node> RandomChoice(const std::vector<std::shared_ptr<Node> > &nodes);
```
Comment (selectionator) [inline cpp_modules/src/mcts/Mcts.h:38]: git delete all :smiling_imp:

File: cpp_modules/src/mcts/Node.h [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+		// Upper confidence bound applied to trees
+		// Q/N + C * (sqrt(log(parent.N)
+		[[nodiscard]] double uct(const double explore) const {
+			return Q / static_cast<float>(N) + explore * sqrt(
+				       log(parent->N));
+		}
+
+	public:
+		// Number of simulations that has been run on this node
```
Comment (selectionator) [inline cpp_modules/src/mcts/Node.h:31]: N and Q from my previous comment

File: cpp_modules/src/tests/MctsTest.cpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+using namespace Sc2::Mcts;
+
+TEST_SUITE("Test MCTS") {
+	TEST_CASE("Can create a Node") {
+		auto node = std::make_shared<Node>(Action::wait, nullptr);
+		CHECK(node->N == 0);
+		CHECK(node->children.size() == 0);
+
+		SUBCASE("Can add children to a node") {
+			// auto childNode = Node(Action::buildBase, std::make_shared<Node>(node));
```
Comment (selectionator) [inline cpp_modules/src/tests/MctsTest.cpp:20]: Git delete all :smiling_imp:

File: cpp_modules/src/tests/MctsTest.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+		}
+	}
+
+
+	TEST_CASE("Can create run an MCTS") {
+		const auto state = std::make_shared<Sc2::State>();
+
+		const auto mcts = new Mcts(state);
+
+		mcts->search(1);
```
Comment (selectionator) [inline cpp_modules/src/tests/MctsTest.cpp:43]: Does this test fail now that we switched to milliseconds?

File: cpp_modules/src/tests/MctsTest.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+		CHECK(bestMove != Action::none);
+	}
+
+	TEST_CASE("Select Node will select a node that has not been fully explored") {
+		const auto state = std::make_shared<Sc2::State>();
+
+		const auto mcts = new Mcts(state);
+
+		mcts->search(1);
```
Comment (selectionator) [inline cpp_modules/src/tests/MctsTest.cpp:55]: Same as before

File: cpp_modules/src/tests/MctsTest.cpp [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+		CHECK(node->children.size() == 0);
+		CHECK(node->getDepth() > 1);
+	}
+
+	TEST_CASE("Expand will expand with all available actions in a state") {
+		const auto state = std::make_shared<Sc2::State>();
+
+		SUBCASE("Will expand to all actions, when there is enough resources") {
+			const auto mcts = new Mcts(state);
+			mcts->search(1);
```
Comment (selectionator) [inline cpp_modules/src/tests/MctsTest.cpp:69]: Same as before

File: cpp_modules/src/tests/Sc2StateTest.cpp [hunk 10]
```diff
@@ -7,225 +7,232 @@
...
-		currentVespene = state->getVespene();
-		auto mineralWorkers = state->getMineralWorkers();
-		auto vespeneWorkers = state->getVespeneWorkers();
-		state->buildBase();
-		currentMinerals = currentMinerals - state->getBuildBaseCost().minerals + (mineralWorkers * MINERAL_PER_WORKER);
-		CHECK(state->getMinerals() == currentMinerals);
-
-		currentVespene = currentVespene - state->getBuildBaseCost().vespene + (vespeneWorkers * VESPENE_PER_WORKER);
-		CHECK(state->getVespene() == currentVespene);
+		state->wait(50);
```
Comment (selectionator) [inline cpp_modules/src/tests/Sc2StateTest.cpp:30]: Same as before (Not just here in this file)

**Review by selectionator — APPROVED** [review]
architecture_AlphaStar

File: cpp_modules/src/mcts/Mcts.cpp (hunk 3 above)
Comment (mkju19) [reply cpp_modules/src/mcts/Mcts.cpp:134]: k

Comment (selectionator) [reply cpp_modules/src/mcts/Mcts.cpp:134]: :(

**Pull request merged**